infrastructure/server-setup/tasks/users.yml

```diff
@@ -0,0 +1,183 @@
+        state: present
+        sudoer: true
+
+- name: Ensure users are present
+  user:
+    name: '{{ item.name }}'
+    state: '{{ item.state }}'
+  with_items: '{{ users }}'
+
+- name: Grant passwordless sudo to the users
+  ansible.builtin.lineinfile:
+    path: /etc/sudoers.d/{{ item.name }}
+    line: '{{ item.name }} ALL=(ALL) NOPASSWD:ALL'
+    validate: 'visudo -cf %s'
+    create: yes
+  become: yes
+  when: item.state == 'present' and item.sudoer == true
+  with_items: '{{ users }}'
+
+- name: Set authorized key for users
+  authorized_key:
+    user: '{{ item.name }}'
+    key: '{{ item.ssh_key }}'
+  when: item.state == 'present'
+  with_items: '{{ users }}'
+
+- name: Ensure user's shell is bash
+  user:
+    name: '{{ item.name }}'
+    shell: /bin/bash
+  when: item.state == 'present'
+  with_items: '{{ users }}'
+
+- name: Ensure TERM is set in .bashrc
+  lineinfile:
+    path: /home/{{ item.name }}/.bashrc
+    line: 'export TERM=xterm'
+    create: yes
+  when: item.state == 'present'
+  with_items: '{{ users }}'
+
+- name: Check if Docker group exists
+  command: getent group docker
+  register: docker_group
+  ignore_errors: yes
+
+- name: Add user to Docker group
+  when: docker_group.rc == 0 and item.state == 'present'
+  with_items: '{{ users }}'
+  ignore_errors: yes
+  user:
+    name: '{{ item.name }}'
+    groups: docker
+    append: yes
+
+- name: Ensure 2FA setup lines are present in .profile for each user
+  blockinfile:
+    path: '/home/{{ item.name }}/.profile'
+    block: |
+      if [ ! -f "$HOME/.google_authenticator" ]; then
+          echo "You must set up two-factor authentication."
+          (trap 'echo "Ctrl+C detected."' INT; google-authenticator) || exit 1
+      fi
+    create: yes
+    owner: '{{ item.name }}'
+    group: '{{ item.name }}'
+    mode: '0644'
+  when: item.state == 'present'
+  with_items: '{{ users }}'
+
+- name: 'Install libpam-google-authenticator'
+  apt:
+    name: libpam-google-authenticator
+    state: present
+    update_cache: yes
+
+- name: Ensure pam_google_authenticator.so is present in /etc/pam.d/sshd
+  lineinfile:
+    path: /etc/pam.d/sshd
+    line: 'auth required pam_google_authenticator.so nullok'
+    state: present
+
+- name: Add condition for users in users.yml for 2FA in /etc/pam.d/sshd
+  lineinfile:
+    path: /etc/pam.d/sshd
+    insertafter: 'auth required pam_google_authenticator.so nullok'
+    line: 'auth required pam_permit.so'
+    state: present
+
+- name: Disable default auth mechanism (password) in PAM
```

Comment:
@rikukissa does this disable password based SSH authentication for all users?

Reply:
Yea, exactly. That removes the default password prompt login for all users. 

More details here [in the Digital Ocean's tutorial](https://www.digitalocean.com/community/tutorials/how-to-set-up-multi-factor-authentication-for-ssh-on-ubuntu-20-04#step-3-making-ssh-aware-of-mfa:~:text=Find%20the%20line%20%40include%20common%2Dauth%20and%20comment%20it%20out%20by%20adding%20a%20%23%20character%20as%20the%20first%20character%20on%20the%20line.%20This%20tells%20PAM%20not%20to%20prompt%20for%20a%20password%3A)